custom_metrics/structured-data.js

```diff
@@ -0,0 +1,144 @@
+  let error = { type: "error", context: context };
+
+  try {
+    if (typeof messageOrException === "string") {
+      error.message = messageOrException;
+    } else if (messageOrException instanceof Object) {
+      error.exception = messageOrException;
+
+      if (messageOrException.message) {
+        error.message = messageOrException.message;
+      } else {
+        error.message = JSON.stringify(
+          messageOrException,
+          Object.getOwnPropertyNames(messageOrException)
+        );
+      }
+    } else {
+      error.message = JSON.stringify(messageOrException);
+    }
+
+    if (exception) {
+      error.exception = exception;
+
+      if (exception.message) {
+        error.message += ": " + exception.message;
+      }
+    }
+  } catch (e) {
+    error.message = "logError failed";
+    error.exception = e;
+  }
+
+  _logs.push(error);
+
+  return error;
+}
+
+var _custom_metrics = null;
+
+try {
+  // whole process is placed in a try/catch so we can log uncaught errors
+
+  // this provides access to a lot of WebPageTest data, including the raw html, headers and other requests involved
+  var _wptBodies = [];
+  try {
+    _wptBodies = $WPT_BODIES;
+  } catch (e) {
+    logError("wptBodies", "Data returned was not valid", e);
+  }
+
+  // string of the raw html
+  let _rawHtml = null;
+  function getRawHtml() {
+    if (!_rawHtml && _wptBodies.length > 0) {
+      _rawHtml = _wptBodies[0].response_body;
+    }
+    return _rawHtml;
+  }
+
+  // an html document of the raw html
+  let _rawHtmlDocument = null;
+  function getRawHtmlDocument() {
+    if (!_rawHtmlDocument) {
+      let html = getRawHtml();
+
+      if (html) {
+        _rawHtmlDocument =
+          document.implementation.createHTMLDocument("New Document");
+
+        _rawHtmlDocument.documentElement.innerHTML = html;
+      }
+    }
+    return _rawHtmlDocument;
+  }
+
+  _custom_metrics = {
+    structured_data: (() => {
+      try {
+        function gatherStructuredData(d) {
+          let target = {
+            jsonld_scripts: [],
+            raw_html: '',
+            meta_tags: []
+          };
+
+          // JSON-LD
+          target.jsonld_scripts = [...d.querySelectorAll('script[type="application/ld+json"i]')].map(script => script.innerHTML)
+
+          // Microdata, RDFa & Microformats
+          target.raw_html = d.documentElement.outerHTML;
+
+          // Dublin Core, Twitter, Facebook & OpenGraph
+          target.meta_tags = [...d.querySelectorAll('meta')].map(tag => {
+            const attributes = tag.getAttributeNames()
+            const json_tag = {}
+            attributes.map(attribute => {
+              json_tag[attribute] = tag.getAttribute(attribute)
+            })
+            return json_tag
+          })
+
+          return target;
+        }
+
+        let result = {};
+        result.rendered = gatherStructuredData(document);
+
+        let rawHtmlDocument = getRawHtmlDocument();
+
+        if (rawHtmlDocument) {
```

Comment:
IIRC the SEO chapter differentiates between raw/rendered markup to get a better sense of dynamically added content. Is that as interesting to the Structured Data chapter, or could this be simplified to look at only the rendered markup?

Reply:
Good point. It might be interesting to report on the way SD is added (dynamic vs static). As the SEO chapter is doing it, might be something worth looking into?

Reply:
I think we'll just leave it in, if that's okay. We'll take a look once we've got the data, and if there's interesting stuff (e.g. some SD being overwritten by injection, or inconsistencies being introduced by injection), then we'll be able to talk about it. Otherwise, we'll just stick with the final rendered page, I imagine.

Reply:
Yep! I don't think this will form a key part of the storytelling, but agreed that it's worth reviewing for any interesting nuances.

Reply:
We've had a bit of a discussion about the Microdata, RDFa & Microformat HTML stuff here: https://httparchive.slack.com/archives/C021GGN9W4D/p1624916290080500.

Again, I think if it's okay, we'll just leave the HTML in, but I'm also just away to add a simple "flag if found" so we can report on `% of pages which include that type of SD`.

We might play around with a HTML parser later, but attempting it in these custom metrics is likely too brittle.

Reply:
I tested this custom metric on the NYTimes website and it produces 2.6 MB of data, in the `raw_html` alone. Keep in mind that we store all of these results in a BigQuery table for millions of pages, so this would become prohibitively large very quickly.

Also note that we already have the raw HTML (see the `response_bodies` dataset) so outputting it here defeats the purpose of using custom metrics, which is to make use of JS APIs at runtime to more efficiently process the document. In short, if we're parsing strings of HTML, we're doing it wrong 😅

Reply:
Removed the `raw_html` part. I think this chapter is focusing on JSON-LD, and we'll be able to talk about general usage of non-JSON-LD with what we have already, without the `raw_html` bit. So we've removed that to save on storage :)